File: sdk/typescript/src/connection.ts
import { homedir } from 'node:os';
import { join } from 'node:path';

import type { Transport } from '@connectrpc/connect';

import { InvalidArgumentError } from './errors.js';

/**
 * Connection configuration for reaching an ArcBox daemon.
 *
 * Resolution order for every field: explicit option > environment > default.
 * The default tier is the local daemon's Unix socket; setting an API URL
 * (option or `ARCBOX_API_URL`) selects the remote tier instead (CORE-63,
 * reserved — no cloud front door exists yet).
 */
export interface ConnectionOptions {
  /** Unix socket path of the local daemon (env: `ARCBOX_SOCKET`). */
  socketPath?: string;
  /** Base URL of a remote daemon / cloud front door (env: `ARCBOX_API_URL`). */
  apiUrl?: string;
  /**
   * Bearer credential attached as an `Authorization` header when set
   * (env: `ARCBOX_API_KEY`). Unused by the local daemon, which trusts
   * socket file permissions instead.
   */
  apiKey?: string;
  /** Per-RPC deadline in milliseconds for unary calls. Streams are exempt. */
  requestTimeoutMs?: number;
  /**
   * Injected connect-es Transport, replacing socket/URL resolution
   * entirely — the mock/testing seam.
   */
  transport?: Transport;
}

/** A fully resolved connection target. */
export interface ResolvedConnection {
  /**
   * Base URL handed to the transport. On the Unix-socket tier this is a
   * placeholder (`http://arcbox`) that only supplies the Host header and
   * request paths — the connection itself goes to `socketPath`.
   */
  baseUrl: string;
  /** Unix socket to dial; unset on the remote (TCP) tier. */
  socketPath?: string;
  /** Bearer credential to attach, when set. */
  apiKey?: string;
  /** Per-unary-RPC deadline in milliseconds, when set. */
  requestTimeoutMs?: number;
}

/**
 * Placeholder authority for Unix-socket requests. Node's `http.request`
 * takes the connection target from `socketPath` and uses the URL only for
 * the Host header and path, so any stable name works here.
 */
export const UDS_BASE_URL = 'http://arcbox';

/** Default socket location relative to the daemon data dir: `run/arcbox.sock`. */
function defaultSocketPath(env: NodeJS.ProcessEnv): string {
  // Mirrors arcbox-constants paths.rs: `<data_dir>/run/arcbox.sock`, data
  // dir from ARCBOX_DATA_DIR, default `~/.arcbox`.
  const dataDir = env.ARCBOX_DATA_DIR ?? join(homedir(), '.arcbox');
  return join(dataDir, 'run', 'arcbox.sock');
}

/**
 * Resolve connection options against the environment.
 *
 * Tier selection: an explicit `socketPath` and an explicit `apiUrl` are
 * contradictory and rejected. At the environment level `ARCBOX_API_URL`
 * wins over `ARCBOX_SOCKET` — setting it selects the remote tier.
 */
export function resolveConnection(
  options: ConnectionOptions = {},
  env: NodeJS.ProcessEnv = process.env,
): ResolvedConnection {
  if (options.socketPath !== undefined && options.apiUrl !== undefined) {
    throw new InvalidArgumentError(
      'connection.socketPath and connection.apiUrl are mutually exclusive: ' +
        'a connection dials either the local Unix socket or a remote URL',
    );
  }

  const apiKey = options.apiKey ?? env.ARCBOX_API_KEY;
  const requestTimeoutMs = options.requestTimeoutMs;
  const common = { apiKey, requestTimeoutMs };

  if (options.socketPath !== undefined) {
    return { baseUrl: UDS_BASE_URL, socketPath: options.socketPath, ...common };
  }
  if (options.apiUrl !== undefined) {
    return { baseUrl: options.apiUrl, ...common };
  }
  if (env.ARCBOX_API_URL !== undefined && env.ARCBOX_API_URL !== '') {
    return { baseUrl: env.ARCBOX_API_URL, ...common };
  }
  const socketPath =
    env.ARCBOX_SOCKET !== undefined && env.ARCBOX_SOCKET !== ''
      ? env.ARCBOX_SOCKET
      : defaultSocketPath(env);
  return { baseUrl: UDS_BASE_URL, socketPath, ...common };
}

File: sdk/rust/src/connection.rs
//! Connecting to the daemon: option and environment resolution.
//!
//! Resolution order: explicit option > environment > default, matching
//! the TypeScript and Python SDKs.
//!
//! | Environment      | Meaning                                             |
//! |------------------|-----------------------------------------------------|
//! | `ARCBOX_SOCKET`  | daemon Unix socket                                  |
//! | `ARCBOX_DATA_DIR`| data dir whose `run/arcbox.sock` is the default     |
//! | `ARCBOX_PROFILE` | `development` switches the default to `~/.arcbox-dev` |
//! | `ARCBOX_API_URL` | remote tier (reserved, CORE-63) — not supported here yet |

use std::env;
use std::path::PathBuf;

use crate::error::{Error, ErrorKind};

/// How to reach the daemon. `Connection::default()` resolves from the
/// environment; [`Connection::socket_path`] overrides it.
#[derive(Debug, Clone, Default)]
pub struct Connection {
    socket_path: Option<PathBuf>,
}

impl Connection {
    /// A connection resolved from the environment (see the module docs).
    #[must_use]
    pub fn new() -> Self {
        Self::default()
    }

    /// Use this daemon socket, ignoring the environment.
    #[must_use]
    pub fn socket_path(mut self, path: impl Into<PathBuf>) -> Self {
        self.socket_path = Some(path.into());
        self
    }

    /// The socket this connection addresses.
    ///
    /// # Errors
    ///
    /// [`ErrorKind::InvalidArgument`] when `ARCBOX_API_URL` selects the
    /// remote tier (this SDK does not speak it yet) or no home directory
    /// exists to anchor the default data dir.
    pub(crate) fn resolve(&self) -> Result<PathBuf, Error> {
        if let Some(path) = &self.socket_path {
            return Ok(path.clone());
        }
        if let Some(socket) = env_non_empty("ARCBOX_SOCKET") {
            return Ok(PathBuf::from(socket));
        }
        if env_non_empty("ARCBOX_API_URL").is_some() {
            return Err(Error::new(
                ErrorKind::InvalidArgument,
                "ARCBOX_API_URL selects the remote tier, which the Rust SDK \
                 does not support yet",
                "connection.resolve",
            )
            .with_suggestion(
                "unset ARCBOX_API_URL or point ARCBOX_SOCKET at a local daemon socket",
            ));
        }
        let data_dir = match env_non_empty("ARCBOX_DATA_DIR") {
            Some(dir) => PathBuf::from(dir),
            None => {
                let home = env_non_empty("HOME").map(PathBuf::from).ok_or_else(|| {
                    Error::new(
                        ErrorKind::InvalidArgument,
                        "no HOME directory to anchor the default ArcBox data dir",
                        "connection.resolve",
                    )
                    .with_suggestion("set ARCBOX_SOCKET or ARCBOX_DATA_DIR explicitly")
                })?;
                let profile_dir = match env_non_empty("ARCBOX_PROFILE").as_deref() {
                    Some("development") => ".arcbox-dev",
                    _ => ".arcbox",
                };
                home.join(profile_dir)
            }
        };
        Ok(data_dir.join("run").join("arcbox.sock"))
    }
}

/// A set, non-empty environment variable — unset and empty are the same
/// "not configured" state, as in the sibling SDKs.
fn env_non_empty(name: &str) -> Option<String> {
    env::var(name).ok().filter(|value| !value.is_empty())
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn explicit_socket_path_wins_over_everything() {
        let connection = Connection::new().socket_path("/tmp/explicit.sock");
        assert_eq!(
            connection.resolve().unwrap(),
            PathBuf::from("/tmp/explicit.sock")
        );
    }
}
